src/aiida_quantumespresso/utils/hubbard.py

```diff
@@ -313,6 +312,243 @@ def get_hubbard_for_supercell(self, supercell: StructureData, thr: float = 1e-3)
 
         return HubbardStructureData.from_structure(structure=supercell, hubbard=new_hubbard)
 
+    def get_interacting_pairs(self) -> Dict[str, List[str]]:
+        """Return tuple of kind name interaction pairs.
+
+        :returns: dictionary of onsite kinds with a list of V kinds
+        """
+        pairs_dict = dict()
+        sites = self.hubbard_structure.sites
+        parameters = self.hubbard_structure.hubbard.parameters
+
+        onsite_indices = [p.atom_index for p in parameters if p.atom_index == p.neighbour_index]
+
+        for index in onsite_indices:
+            name = sites[index].kind_name
+            if name not in pairs_dict:
+                pairs_dict[name] = []
+
+        for parameter in parameters:
+            onsite_name = sites[parameter.atom_index].kind_name
+            neigh_name = sites[parameter.neighbour_index].kind_name
+
+            if onsite_name in pairs_dict and onsite_name != neigh_name:
+                if not neigh_name in pairs_dict[onsite_name]:
+                    pairs_dict[onsite_name].append(neigh_name)
+
+        return pairs_dict
+
+    def get_pairs_radius(
+        self,
+        onsite_index: int,
+        neighbours_names: List[str],
+        number_of_neighbours: int,
+        radius_max: float = 7.0,
+        thr: float = 1.0e-2,
+    ) -> Tuple[float, float]:
+        """Return the minimum and maximum radius of the first neighbours of the onsite site.
+
+        :param onsite_index: index in the structure of the onsite Hubbard atom
+        :param neighbours_names: kind names of the neighbours
+        :param number_of_neighbours: number of neighbours coming to select
+        :param radius_max: maximum radius (in Angstrom) to use for looking for neighbours
+        :param thr: threshold (in Angstrom) for defining the shells
+        :return: (radius min +thr, radius max -thr) defining the shells containing only the first neighbours
+        """
+        rmin = 0
+        pymat = self.hubbard_structure.get_pymatgen_structure()
```

Comment:
Yeah, good point, we can do that, I don't think there should be something incompatible in doing so. We just need to open an issue so we don't forget to make the code cleaner. Or actually, we can even think to override within this PR the `get_pymathen_structure`?

Reply:
you could in principle create the sites as PeriodicSite

```
sites = [
            PeriodicSite(
                species=site.species,
                coords=site.coords,
                lattice=Lattice(self.hubbard_structure.cell, pbc=self.hubbard_structure.pbc),
                coords_are_cartesian=True
            ) for site in self.hubbard_structure.get_pymatgen().sites
        ]
```

And then instead of using the pymat StructureData , just call the function from pymatgen, that it accepts a list of periodic sites ?

get_neighbor_list(r: float, sites: Sequence[[PeriodicSite](https://pymatgen.org/pymatgen.core.html#pymatgen.core.sites.PeriodicSite)] | None = None, numerical_tol: float = 1e-08, exclude_self: bool = True)→ tuple[np.ndarray, ...][[source]](https://github.com/materialsproject/pymatgen/blob/v2024.3.1/pymatgen/core/structure.py#L1615-L1672)[](https://pymatgen.org/pymatgen.core.html#pymatgen.core.structure.IStructure.get_neighbor_list)